### ISSUE TITLE
Mypy reports incompatible types for an optional dataclass field with generic default factory

### ISSUE THREAD

**Issue by sagittarian** [python/mypy#11923]
**Bug Report**

In a dataclass with an optional field with a generic as the default factory, mypy reports that the types are incompatible when there should be no compatibility issue between them.

```python
from typing import List, Optional
from dataclasses import dataclass, field


@dataclass
class A:
    x: Optional[List[int]] = field(default_factory=list)
```

**Expected Behavior**

A bare list is compatible with `Optional[List[int]]`, so no error should be reported.

**Actual Behavior**

```
$ mypy ./optlist.py 
optlist.py:7: error: Incompatible types in assignment (expression has type "List[_T]", variable has type "Optional[List[int]]")
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.3 LTS"
$ python --version
Python 3.8.10
$ mypy --version
mypy 0.930
```

**Comment by yoann9344** [comment]
```python
def optional_list() -> list | None:
    return list()


@dataclass
class A:
    x: Optional[List[int]] = field(default_factory=optional_list)
```

**Comment by Pycz** [comment]
Another workaround:

```python
from typing import Dict, Any, cast
from dataclasses import dataclass, field


@dataclasses.dataclass
class A:
    data: Dict[str, Any] = dataclasses.field(default_factory=cast(Dict[str, Any], dict))
```

**Comment by liav-teichner** [comment]
I think the workaround above has a mistake in it - it tries to cast a callable into a dictionary. The correct one is:

```python
from typing import Dict, List, Optional, Tuple, Any, cast, Callable
from dataclasses import dataclass, field

@dataclass
class A:
    data: Dict[str, Any] = field(default_factory=cast(Callable[[], Dict[str, Any]], dict))
```

**Comment by ilevkivskyi** [comment]
The original examples passes without errors on current master.
